### ISSUE TITLE
tsdb: Head compaction/snapshot inefficiency

### ISSUE THREAD

**Issue by codesome** [prometheus/prometheus#5863]
During snapshot, there are chances of head chunks going beyond the block meta during compaction as we already know and have a fix for. In such case, we add a tombstone to `dranges` here https://github.com/prometheus/prometheus/blob/8755434350ba9f8d3e5434826a9fcfb3a1327cb0/tsdb/compact.go#L766

But it is possible for every series to encounter this. Hence the `dranges` will grow linearly with every series iteration and with the same (duplicate) drange (million series = min of million dranges for the last series!). Memory consumption and snapshot time might increase with this.

**Comment by gerardo8a** [comment]
I'm still on debugging mode, but I have two prometheus deployments in different colos, they are all similar (still trying to find out any difference) but one of the deployments memory starts increasing and it never goes down until the pod is OOMKilled. 

I had some issues previously and this was related to some custom annotations (long ones) being sent to prometheus and also removed some other metrics to prevent this from happening. This fixed my issue, but as described here it seems to be some sort of cycle where memory just keeps growing and its never released 

Bad deployment
![Screen Shot 2020-05-12 at 7 17 41 PM](https://user-images.githubusercontent.com/217191/81764055-8029b680-9485-11ea-85fb-e25dea9a23f5.png)

Similar deployment no issues
<img width="1577" alt="Screen Shot 2020-05-12 at 7 18 44 PM" src="https://user-images.githubusercontent.com/217191/81764073-89b31e80-9485-11ea-88c1-e77e4605675b.png">

**Comment by brian-brazil** [comment]
@codesome Is this still relevant, especially given isolation and recent changes to tombstone handling in head?

**Comment by codesome** [comment]
Yes, it is still relevant. `dranges` can still get a lot of duplicate ranges as mentioned in the issue description. This should be an easy fix with a boolean I guess.

**Comment by codesome** [comment]
My bad! This issue was not relevant since the beginning
https://github.com/prometheus/prometheus/blob/8755434350ba9f8d3e5434826a9fcfb3a1327cb0/tsdb/compact.go#L742
`dranges` is always a fresh slice for every series, so, that particular append will happen only once.

@gerardo8a Your issue seems to be different than the one mentioned here. If you are still facing it and if you think it is not an issue in config or other settings, can you open another issue with more details?
